wallet/invoices.c

```diff
@@ -0,0 +1,402 @@
+		sqlite3_bind_int64(stmt, 4, *msatoshi);
+	else
+		sqlite3_bind_null(stmt, 4);
+	sqlite3_bind_text(stmt, 5, label, strlen(label), SQLITE_TRANSIENT);
+	sqlite3_bind_int64(stmt, 6, expiry_time);
+
+	db_exec_prepared(invoices->db, stmt);
+
+	/* Create and load in-memory structure. */
+	invoice	= tal(invoices, struct invoice);
+
+	invoice->id = sqlite3_last_insert_rowid(invoices->db->sql);
+	invoice->state = UNPAID;
+	invoice->label = tal_strdup(invoice, label);
+	invoice->msatoshi = tal_dup(invoice, u64, msatoshi); /* Works even if msatoshi == NULL. */
+	memcpy(&invoice->r, &r, sizeof(invoice->r));
+	memcpy(&invoice->rhash, &rhash, sizeof(invoice->rhash));
+	invoice->expiry_time = expiry_time;
+	invoice->pay_index = 0;
+	list_head_init(&invoice->waitone_waiters);
+
+	/* Add to invoices object. */
+	list_add_tail(&invoices->invlist, &invoice->list);
+
+	return invoice;
+}
+
+
+const struct invoice *invoices_find_by_label(struct invoices *invoices,
+					     const char *label)
+{
+	struct invoice *i;
+
+	/* FIXME: Use something better than a linear scan. */
+	list_for_each(&invoices->invlist, i, list) {
+		if (streq(i->label, label))
+			return i;
+	}
+	return NULL;
+}
+
+const struct invoice *invoices_find_unpaid(struct invoices *invoices,
+					   const struct sha256 *rhash)
+{
+	struct invoice *i;
+
+	list_for_each(&invoices->invlist, i, list) {
+		if (structeq(rhash, &i->rhash) && i->state == UNPAID) {
+			if (time_now().ts.tv_sec > i->expiry_time)
+				break;
+			return i;
+		}
+	}
+	return NULL;
+}
+
+bool invoices_delete(struct invoices *invoices,
+		     const struct invoice *cinvoice)
+{
+	sqlite3_stmt *stmt;
+	struct invoice_waiter *w;
+	struct invoice *invoice = (struct invoice *) cinvoice;
+	const tal_t *tmpctx = tal_tmpctx(NULL);
+
+	/* Delete from database. */
+	stmt = db_prepare(invoices->db, "DELETE FROM invoices WHERE id=?;");
+	sqlite3_bind_int64(stmt, 1, invoice->id);
+	db_exec_prepared(invoices->db, stmt);
+
+	if (sqlite3_changes(invoices->db->sql) != 1)
+		return false;
+
+	/* Delete from invoices object. */
+	list_del_from(&invoices->invlist, &invoice->list);
+
+	/* Tell all the waiters about the fact that it was deleted. */
+	while ((w = list_pop(&invoice->waitone_waiters,
+			     struct invoice_waiter,
+			     list)) != NULL) {
+		/* Acquire the watcher for ourself first. */
+		tal_steal(tmpctx, w);
+		trigger_invoice_waiter(w, NULL);
+	}
+
+	/* Free all watchers and the invoice. */
+	tal_free(tmpctx);
+	tal_free(invoice);
+	return true;
+}
+
+const struct invoice *invoices_iterate(struct invoices *invoices,
+				       const struct invoice *invoice)
+{
+	if (invoice)
+		return list_next(&invoices->invlist, invoice, list);
+	else
+		return list_top(&invoices->invlist, struct invoice, list);
+}
+
+static s64 get_next_pay_index(struct db *db)
+{
+	/* Equivalent to (next_pay_index++) */
+	s64 next_pay_index;
+	next_pay_index = db_get_intvar(db, "next_pay_index", 0);
+	/* Variable should exist. */
+	assert(next_pay_index > 0);
```

Comment:
Ok, took me a while to find the corresponding migration, but this is ok :+1:

Reply:
\^\^